src/git.md

```diff
@@ -0,0 +1,187 @@
+encounter at least one "rebase conflict." This happens when git's attempt to
+reapply the changes fails because your changes conflicted with other changes
+that have been made. You can tell that this happened because you'll see
+lines in the output that look like
+
+```
+CONFLICT (content): Merge conflict in file.rs
+```
+
+When you open these files, you'll see sections of the form
+
+```
+<<<<<<< HEAD
+Original code
+=======
+Your code
+>>>>>>> 8fbf656... Commit fixes 12345
+```
+
+This represents the lines in the file that git could not figure out how to
+rebase. The section between `<<<<<<< HEAD` and `=======` has the code from
+master, while the other side has your version of the code. You'll need to
+decide how to deal with the conflict. You may want to keep your changes,
+keep the changes on master, or combine the two.
+ 
+Generally, resovling the conflict consists of two steps: First, fix the
+particular conflict. Edit the file to make the changes you want and remove the
+`<<<<<<<`, `=======` and `>>>>>>>` lines in the process. Second, check the
+surrounding code. If there was a conflict, its because someone else changed the
+same code you did. That means its likely there are some logical errors lying
+around too!
+
+Once you're all done fixing the conflicts, you need to stage the files that had
+conflicts in them via `git add`. Afterwards, run `git rebase --continue` to let
+git know that you've resolved the conflicts and it should finish the rebase.
+Once the rebase has succeeded, you'll want to update the associated branch on
+your fork with `git push -f`.
+
+Note that `git push` will not work properly and say something like this:
+
+```
+ ! [rejected]        issue-xxxxx -> issue-xxxxx (non-fast-forward)
+error: failed to push some refs to 'https://github.com/username/rust.git'
+hint: Updates were rejected because the tip of your current branch is behind
+hint: its remote counterpart. Integrate the remote changes (e.g.
+hint: 'git pull ...') before pushing again.
+hint: See the 'Note about fast-forwards' in 'git push --help' for details.
+```
+
+The advice this gives is incorrect! Because of the "no-merge" policy, running
+`git pull` will create a merge commit, defeating the point of your rebase. Use
+`git push -f` instead.
+
+## Advanced Rebasing
+
+Sometimes, you may want to perform a more complicated rebase. There are two
+common scenarios that might call for this.
+
+If your branch contains multiple consecutive rewrites of the same code, or if
+the rebase conflicts are extremely severe, it is possible that just trying to
+reapply the changes you made on top of the updated code will be too much of a
+headache. In this case, you can use the interactive rebase feature via
+`git rebase -i master` to gain more control over the process. This allows you
+to choose to skip commits because they represent changes you no longer need,
+edit the commits that you do not skip, or change the order in which they are
+applied.
```

Comment:
If things are normally this bad, I normally don't use rebasing at all. Usually this happens because I made a _ton_ of changes or because I unindented a function and git thinks every line conflicts.

For the first one, I give up on keeping the commit history and do something like this:

```
git checkout feature-branch
# squash all the changes into one commit so you only have to worry about conflicts once, not for every commit
git rebase -i $(git merge-base master)  # and squash all changes along the way
git rebase master 
# fix all merge conflicts
git rebase --continue
```

That throws away the history but means you only have to solve conflicts once instead of again and again for every commit. See https://github.com/rust-lang/rust/pull/75756 or https://github.com/rust-lang/rust/pull/74489 for examples of big PRs like this that are hard to rebase.

For the second one, there's usually not many conflicts if you say to ignore whitespace changes: https://stackoverflow.com/questions/24412980/how-to-run-git-rebase-and-ignore-whitespace
That messes up the formatting so you have to run `x.py fmt`, but that's way easier than fixing merge conflicts.